File: packages/oxlint-plugin-react-doctor/src/plugin/rules/react-builtins/no-string-refs.ts
import { defineRule } from "../../utils/define-rule.js";
import type { EsTreeNode } from "../../utils/es-tree-node.js";
import type { EsTreeNodeOfType } from "../../utils/es-tree-node-of-type.js";
import { getParentComponent } from "../../utils/get-parent-component.js";
import { isNodeOfType } from "../../utils/is-node-of-type.js";
import { isTestlikeFilename } from "../../utils/is-testlike-filename.js";
import type { Rule } from "../../utils/rule.js";

const STRING_IN_REF_MESSAGE =
  "Your component can't reach this node because string refs don't work in modern React.";
const THIS_REFS_MESSAGE =
  "Your component can't reach its nodes because `this.refs` is empty in modern React.";

const isStringLiteralRefAttribute = (
  attribute: EsTreeNodeOfType<"JSXAttribute">,
  noTemplateLiterals: boolean,
): boolean => {
  if (!isNodeOfType(attribute.name, "JSXIdentifier") || attribute.name.name !== "ref") {
    return false;
  }
  const value = attribute.value;
  if (!value) return false;
  if (isNodeOfType(value, "Literal") && typeof value.value === "string") return true;
  if (isNodeOfType(value, "JSXExpressionContainer")) {
    const expression = value.expression;
    if (isNodeOfType(expression, "Literal") && typeof expression.value === "string") return true;
    if (noTemplateLiterals && isNodeOfType(expression, "TemplateLiteral")) return true;
  }
  return false;
};

interface NoStringRefsSettings {
  noTemplateLiterals?: boolean;
}

const resolveSettings = (
  settings: Readonly<Record<string, unknown>> | undefined,
): NoStringRefsSettings => {
  const reactDoctorSettings = settings?.["react-doctor"];
  if (
    typeof reactDoctorSettings === "object" &&
    reactDoctorSettings !== null &&
    "noStringRefs" in reactDoctorSettings &&
    typeof (reactDoctorSettings as { noStringRefs?: unknown }).noStringRefs === "object" &&
    (reactDoctorSettings as { noStringRefs?: unknown }).noStringRefs !== null
  ) {
    const inner = (reactDoctorSettings as { noStringRefs: NoStringRefsSettings }).noStringRefs;
    return inner;
  }
  return {};
};

// Port of `oxc_linter::rules::react::no_string_refs`. Reports two shapes:
//   1. `<x ref="..." />` or `<x ref={"..."} />` — string-literal refs.
//      With `noTemplateLiterals: true` setting, also reports template
//      literals (`<x ref={\`...\`} />`).
//   2. `this.refs.foo` — when used inside a React component's body.
export const noStringRefs = defineRule<Rule>({
  id: "no-string-refs",
  title: "Use of string refs",
  severity: "warn",
  recommendation:
    "Use a callback ref (`ref={(node) => { this.foo = node }}`) or `useRef` instead of string refs.",
  create: (context) => {
    const { noTemplateLiterals = false } = resolveSettings(context.settings);
    // Test files routinely use library-specific JSX DSLs whose `ref`
    // attribute means something other than React's deprecated string
    // refs (e.g. tldraw's `createShapesFromJsx([<TL.geo ref="boxA"/>])`
    // where the string identifies a shape in the test, not a React ref).
    // Skip those — string-ref-deprecation warnings only matter for
    // production React code, where the test corpus showed zero hits.
    const isTestlikeFile = isTestlikeFilename(context.filename);
    return {
      JSXAttribute(node: EsTreeNodeOfType<"JSXAttribute">) {
        if (isTestlikeFile) return;
        if (isStringLiteralRefAttribute(node, noTemplateLiterals)) {
          context.report({ node, message: STRING_IN_REF_MESSAGE });
        }
      },
      MemberExpression(node: EsTreeNodeOfType<"MemberExpression">) {
        if (isTestlikeFile) return;
        if (!isNodeOfType(node.object, "ThisExpression")) return;
        if (!isNodeOfType(node.property, "Identifier") || node.property.name !== "refs") return;
        const enclosingComponent: EsTreeNode | null = getParentComponent(node);
        if (!enclosingComponent) return;
        context.report({ node, message: THIS_REFS_MESSAGE });
      },
    };
  },
});

File: native/oxlint/rules/no-string-refs.rs
use oxc_ast::{
    AstKind, MemberExpressionKind,
    ast::{
        Expression, JSXAttribute, JSXAttributeName, JSXAttributeValue, JSXExpression,
    },
};
use oxc_diagnostics::OxcDiagnostic;
use oxc_macros::declare_oxc_lint;
use oxc_span::GetSpan;

use crate::{
    AstNode,
    context::{ContextHost, LintContext},
    rule::Rule,
    utils::get_parent_component,
};

const STRING_IN_REF_MESSAGE: &str =
    "Your component can't reach this node because string refs don't work in modern React.";
const THIS_REFS_MESSAGE: &str =
    "Your component can't reach its nodes because `this.refs` is empty in modern React.";

#[derive(Debug, Default, Clone)]
pub struct NoStringRefs;

declare_oxc_lint!(
    /// Disallow legacy string refs and this.refs access.
    NoStringRefs,
    react_doctor_native,
    correctness,
    version = "0.1.0",
    short_description = "Disallow legacy React string refs.",
);

impl Rule for NoStringRefs {
    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
        let report = match node.kind() {
            AstKind::JSXAttribute(attribute)
                if is_string_literal_ref_attribute(attribute, no_template_literals(ctx)) =>
            {
                Some((STRING_IN_REF_MESSAGE, attribute.span))
            }
            member_kind if member_kind.is_member_expression_kind() => {
                let Some(member_expression) = member_kind.as_member_expression_kind() else {
                    return;
                };
                if is_this_refs(member_expression) && get_parent_component(node, ctx).is_some() {
                    Some((THIS_REFS_MESSAGE, member_expression.span()))
                } else {
                    None
                }
            }
            _ => None,
        };
        let Some((message, span)) = report else {
            return;
        };
        ctx.diagnostic(OxcDiagnostic::warn(message).with_label(span));
    }

    fn should_run(&self, ctx: &ContextHost) -> bool {
        !is_non_production_file(ctx)
    }
}

fn is_string_literal_ref_attribute(attribute: &JSXAttribute, no_templates: bool) -> bool {
    if !matches!(
        &attribute.name,
        JSXAttributeName::Identifier(identifier) if identifier.name == "ref"
    ) {
        return false;
    }
    match &attribute.value {
        Some(JSXAttributeValue::StringLiteral(_)) => true,
        Some(JSXAttributeValue::ExpressionContainer(container)) => {
            matches!(container.expression, JSXExpression::StringLiteral(_))
                || (no_templates
                    && matches!(container.expression, JSXExpression::TemplateLiteral(_)))
        }
        _ => false,
    }
}

fn is_this_refs(member_expression: MemberExpressionKind) -> bool {
    let (object, has_refs_property) = match member_expression {
        MemberExpressionKind::Static(member_expression) => (
            &member_expression.object,
            member_expression.property.name == "refs",
        ),
        MemberExpressionKind::Computed(member_expression) => (
            &member_expression.object,
            matches!(
                &member_expression.expression,
                Expression::Identifier(identifier) if identifier.name == "refs"
            ),
        ),
        MemberExpressionKind::PrivateField(_) => return false,
    };
    has_refs_property && matches!(object, Expression::ThisExpression(_))
}

fn no_template_literals(ctx: &LintContext) -> bool {
    ctx.settings()
        .json
        .as_ref()
        .and_then(|settings| settings.get("react-doctor"))
        .and_then(serde_json::Value::as_object)
        .and_then(|settings| settings.get("noStringRefs"))
        .and_then(serde_json::Value::as_object)
        .and_then(|settings| settings.get("noTemplateLiterals"))
        .and_then(serde_json::Value::as_bool)
        .unwrap_or(false)
}
